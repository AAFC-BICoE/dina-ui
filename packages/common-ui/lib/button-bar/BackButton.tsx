import Link from "next/link";
import React from "react";
import { CommonMessage } from "../intl/common-ui-intl";
import { useBackButtonReloadLastSearch } from "../list-page/reload-last-search/useBackButtonReloadLastSearch";

interface BackButtonProps {
  /** If the id is set, it will return to the view. If it's not, it will return to the list. */
  entityId?: string;

  /** The link type for where to redirect the user. Gets appended with "/" + entityLink + "/edit/". */
  entityLink: string;

  /** Boolean to use to by pass the view page if there is none */
  byPassView?: boolean;

  className?: string;
  buttonMsg?: string;

  /** If going back to a query page, enable to to reload the last performed query. */
  reloadLastSearch?: boolean;
}

/**
 * Cancel Button which is commonly used in the button bar.
 */
export function BackButton({
  entityId,
  entityLink,
  byPassView,
  className,
  buttonMsg,
  reloadLastSearch
}: BackButtonProps) {
  // If reload last search is activated, intercept the back button of the browser to inject the param.
  useBackButtonReloadLastSearch(!reloadLastSearch);

  // When editing an existing entity, the link points to the entity details page.
  // When editing a new entity, the link points to the list page.
  // When placed in view page, will accept url to navigate to
  const { href, message } =
    byPassView || !entityId
      ? {
          href:
            entityLink +
            "/list" +
            (reloadLastSearch ? "?reloadLastSearch" : ""),
          message: "backToList" as const
        }
      : {
          href: `${entityLink}/view?id=${entityId}`,
          message: "backToReadOnlyPage" as const
        };

  return (
    <Link href={href}>
      <a className={`back-button my-auto ${className ? className : ""}`}>
        <CommonMessage id={(buttonMsg as any) ?? message} />
      </a>
    </Link>
  );
}
