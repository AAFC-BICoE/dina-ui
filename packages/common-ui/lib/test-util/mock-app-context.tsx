import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { useMemo, useRef } from "react";
import { SWRConfig } from "swr";
import { PartialDeep } from "type-fest";
import { AccountContextI, AccountProvider } from "../account/AccountProvider";
import {
  ApiClientI,
  ApiClientImpl,
  ApiClientProvider
} from "../api-client/ApiClientContext";
import { CommonUIIntlProvider } from "../intl/common-ui-intl";
import { ModalProvider } from "../modal/modal";

interface MockAppContextProviderProps {
  apiContext?: PartialDeep<ApiClientI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext = { apiClient: { get: () => undefined as any } },
  children
}: MockAppContextProviderProps) {
  const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = useMemo(
    () => ({
      authenticated: true,
      groupNames: ["aafc", "cnc"],
      initialized: true,
      login: noop,
      logout: noop,
      roles: ["user"],
      token: "test-token",
      username: "test-user",
      isAdmin: false
    }),
    []
  );

  const DEFAULT_API_CONTEXT_VALUE = useMemo(
    () =>
      new ApiClientImpl({
        newId: () => "00000000-0000-0000-0000-000000000000"
      }),
    []
  );

  const apiContextWithWarnings = {
    ...apiContext,
    // Add a warning when bulkGet doesn't return anything in a test:
    bulkGet: async (paths: string[], options) => {
      const resources = await apiContext?.bulkGet?.(paths, options);
      if (!resources) {
        console.warn("No response returned for bulkGet paths: ", paths);
      } else {
        resources.forEach((resource, index) => {
          if (!resource) {
            console.warn("No value returned for bulkGet path: ", paths[index]);
          }
        });
      }
      return resources;
    }
  };

  const modalWrapperRef = useRef<HTMLDivElement>(null);

  return (
    <SWRConfig
      // Reset SWR cache between tests.
      value={{ provider: () => new Map() }}
    >
      <AccountProvider
        value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
      >
        <ApiClientProvider
          value={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContextWithWarnings)}
        >
          <CommonUIIntlProvider>
            <div ref={modalWrapperRef}>
              <ModalProvider appElement={modalWrapperRef.current}>
                {children}
              </ModalProvider>
            </div>
          </CommonUIIntlProvider>
        </ApiClientProvider>
      </AccountProvider>
    </SWRConfig>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
