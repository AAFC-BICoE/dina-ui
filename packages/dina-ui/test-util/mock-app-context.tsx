import {
  AccountContextI,
  AccountProvider,
  ApiClientI,
  ApiClientImpl,
  ApiClientProvider,
  ModalProvider
} from "common-ui";
import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { useMemo, useRef } from "react";
import { DndProvider } from "react-dnd-cjs";
import HTML5Backend from "react-dnd-html5-backend-cjs";
import { SWRConfig } from "swr";
import { PartialDeep } from "type-fest";
import { FileUploadProviderImpl } from "../components/object-store/file-upload/FileUploadProvider";
import { DinaIntlProvider } from "../intl/dina-ui-intl";

interface MockAppContextProviderProps {
  apiContext?: PartialDeep<ApiClientI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext = { apiClient: { get: () => undefined as any } },
  children
}: MockAppContextProviderProps) {
  const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = useMemo(
    () => ({
      authenticated: true,
      groupNames: ["aafc", "cnc"],
      initialized: true,
      login: noop,
      logout: noop,
      roles: ["user"],
      token: "test-token",
      username: "test-user",
      isAdmin: false
    }),
    []
  );

  const DEFAULT_API_CONTEXT_VALUE = useMemo(
    () =>
      new ApiClientImpl({
        newId: () => "00000000-0000-0000-0000-000000000000"
      }),
    []
  );

  const apiContextWithWarnings = {
    ...apiContext,
    // Add a warning when bulkGet doesn't return anything in a test:
    bulkGet: async (paths: string[], options) => {
      const resources = await apiContext?.bulkGet?.(paths, options);
      if (!resources) {
        console.warn("No response returned for bulkGet paths: ", paths);
      } else {
        resources.forEach((resource, index) => {
          if (!resource) {
            console.warn("No value returned for bulkGet path: ", paths[index]);
          }
        });
      }
      return resources;
    }
  };

  const modalWrapperRef = useRef<HTMLDivElement>(null);

  return (
    <SWRConfig
      // Reset SWR cache between tests.
      value={{ provider: () => new Map() }}
    >
      <AccountProvider
        value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
      >
        <ApiClientProvider
          value={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContextWithWarnings)}
        >
          <FileUploadProviderImpl>
            <DinaIntlProvider>
              <DndProvider backend={HTML5Backend}>
                <div ref={modalWrapperRef}>
                  <ModalProvider appElement={modalWrapperRef.current}>
                    {children}
                  </ModalProvider>
                </div>
              </DndProvider>
            </DinaIntlProvider>
          </FileUploadProviderImpl>
        </ApiClientProvider>
      </AccountProvider>
    </SWRConfig>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
