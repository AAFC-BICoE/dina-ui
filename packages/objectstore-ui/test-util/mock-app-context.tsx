import {
  AccountContextI,
  AccountProvider,
  ApiClientContextI,
  AuthenticatedApiClientProvider,
  createContextValue,
  ModalProvider
} from "common-ui";
import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { ObjectStoreIntlProvider } from "../intl/objectstore-intl";

interface MockAppContextProviderProps {
  apiContext?: Partial<ApiClientContextI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext,
  children
}: MockAppContextProviderProps) {
  return (
    <AccountProvider
      value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
    >
      <AuthenticatedApiClientProvider
        apiContext={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContext)}
      >
        <ObjectStoreIntlProvider>
          <ModalProvider appElement={document.querySelector("body")}>
            {children}
          </ModalProvider>
        </ObjectStoreIntlProvider>
      </AuthenticatedApiClientProvider>
    </AccountProvider>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}

const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = {
  authenticated: true,
  groups: ["/aafc", "cnc"],
  initialized: true,
  login: noop,
  logout: noop,
  token: "test-token",
  username: "test-user"
};

const DEFAULT_API_CONTEXT_VALUE = createContextValue({
  getTempIdGenerator: () => () => "00000000-0000-0000-0000-000000000000"
});
