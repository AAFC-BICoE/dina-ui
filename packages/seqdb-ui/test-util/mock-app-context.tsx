import {
  ApiClientContext,
  ApiClientContextI,
  createContextValue
} from "common-ui";
import { mount } from "enzyme";
import { SeqdbIntlProvider } from "../intl/seqdb-intl";

interface MockAppContextProviderProps {
  apiContext?: ApiClientContextI;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  apiContext,
  children
}: MockAppContextProviderProps) {
  return (
    <ApiClientContext.Provider value={apiContext || createContextValue()}>
      <SeqdbIntlProvider>{children}</SeqdbIntlProvider>
    </ApiClientContext.Provider>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
