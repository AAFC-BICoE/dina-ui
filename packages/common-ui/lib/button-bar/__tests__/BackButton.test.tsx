import { mountWithAppContext } from "../../test-util/mock-app-context";
import { BackButton } from "../BackButton";

const ENTITY_ID = "108559ed-e000-49c4-95e0-03dee7bfce9b";
const ENTITY_LINK = "/collection/material-sample";

describe("BackButton", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("Back to entity button", () => {
    // Simple back to list button snapshot test.
    const backToEntityButtonWrapper = mountWithAppContext(
      <BackButton entityLink={ENTITY_LINK} entityId={ENTITY_ID} />
    );

    expect(backToEntityButtonWrapper.html()).toMatchSnapshot();

    // Try adding a custom className to ensure it's added.
    const classNameTest = mountWithAppContext(
      <BackButton
        entityLink={ENTITY_LINK}
        entityId={ENTITY_ID}
        className="class-name-test"
      />
    );

    expect(classNameTest.html()).toMatchSnapshot();
  });

  it("Back to list button", () => {
    // Simple back to list button snapshot test.
    const backToListButtonWrapper = mountWithAppContext(
      <BackButton entityLink={ENTITY_LINK} reloadLastSearch={true} />
    );

    expect(backToListButtonWrapper.html()).toMatchSnapshot();

    // Back to entity but with bypass view option enabled, should point to list.
    const backToEntityButtonWrapper = mountWithAppContext(
      <BackButton
        entityLink={ENTITY_LINK}
        entityId={ENTITY_ID}
        byPassView={true}
      />
    );

    expect(backToEntityButtonWrapper.html()).toMatchSnapshot();

    // Test with reload last search option snapshot test.
    const reloadLastSearchWrapper = mountWithAppContext(
      <BackButton entityLink={ENTITY_LINK} reloadLastSearch={true} />
    );

    expect(reloadLastSearchWrapper.html()).toMatchSnapshot();
  });
});
