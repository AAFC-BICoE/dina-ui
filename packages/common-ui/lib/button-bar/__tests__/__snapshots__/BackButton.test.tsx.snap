// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackButton Back to entity button 1`] = `"<div><a class=\\"back-button my-auto \\" href=\\"/collection/material-sample/view?id=108559ed-e000-49c4-95e0-03dee7bfce9b\\">Back to read-only page</a></div>"`;

exports[`BackButton Back to entity button 2`] = `"<div><a class=\\"back-button my-auto class-name-test\\" href=\\"/collection/material-sample/view?id=108559ed-e000-49c4-95e0-03dee7bfce9b\\">Back to read-only page</a></div>"`;

exports[`BackButton Back to list button 1`] = `"<div><a class=\\"back-button my-auto \\" href=\\"/collection/material-sample/list?reloadLastSearch\\">Back to list page</a></div>"`;

exports[`BackButton Back to list button 2`] = `"<div><a class=\\"back-button my-auto \\" href=\\"/collection/material-sample/list\\">Back to list page</a></div>"`;

exports[`BackButton Back to list button 3`] = `"<div><a class=\\"back-button my-auto \\" href=\\"/collection/material-sample/list?reloadLastSearch\\">Back to list page</a></div>"`;
